pkg/resources/database_role.go

```diff
@@ -0,0 +1,211 @@
+	"encoding/csv"
+	"errors"
+	"fmt"
+	"log"
+	"strings"
+
+	"github.com/Snowflake-Labs/terraform-provider-snowflake/pkg/snowflake"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+)
+
+const (
+	databaseRoleIDDelimiter = '|'
+)
+
+var databaseRoleSchema = map[string]*schema.Schema{
+	"name": {
+		Type:        schema.TypeString,
+		Required:    true,
+		Description: "Specifies the identifier for the database role.",
+		ForceNew:    true,
+	},
+	"database": {
+		Type:        schema.TypeString,
+		Required:    true,
+		Description: "The database in which to create the database role.",
+		ForceNew:    true,
+	},
+	"comment": {
+		Type:        schema.TypeString,
+		Optional:    true,
+		Description: "Specifies a comment for the database role.",
+	},
+}
+
+type dbRoleID struct {
+	DatabaseName string
+	RoleName     string
+}
+
+// String() takes in a dbRoleID object and returns a pipe-delimited string:
+// DatabaseName|RoleName.
+func (id *dbRoleID) String() (string, error) {
+	var buf bytes.Buffer
+	csvWriter := csv.NewWriter(&buf)
+	csvWriter.Comma = databaseRoleIDDelimiter
+	dataIdentifiers := [][]string{{id.DatabaseName, id.RoleName}}
+	if err := csvWriter.WriteAll(dataIdentifiers); err != nil {
+		return "", err
+	}
+	strDbRoleID := strings.TrimSpace(buf.String())
+	return strDbRoleID, nil
+}
+
+// databaseRoleIDFromString() takes in a pipe-delimited string: DatabaseName|RoleName
+// and returns a dbRoleID object.
+func databaseRoleIDFromString(stringID string) (*dbRoleID, error) {
+	reader := csv.NewReader(strings.NewReader(stringID))
+	reader.Comma = pipeIDDelimiter
+	lines, err := reader.ReadAll()
+	if err != nil {
+		return nil, fmt.Errorf("not CSV compatible")
+	}
+	if len(lines) != 1 {
+		return nil, fmt.Errorf("1 line per database role")
+	}
+	if len(lines[0]) != 2 {
+		return nil, fmt.Errorf("2 fields allowed")
+	}
+
+	dbRoleResult := &dbRoleID{
+		DatabaseName: lines[0][0],
+		RoleName:     lines[0][1],
+	}
+	return dbRoleResult, nil
+}
+
+// DatabaseRole returns a pointer to the resource representing a database role.
+func DatabaseRole() *schema.Resource {
+	return &schema.Resource{
+		Create: CreateDatabaseRole,
+		Read:   ReadDatabaseRole,
+		Update: UpdateDatabaseRole,
+		Delete: DeleteDatabaseRole,
+
+		Schema: databaseRoleSchema,
+		Importer: &schema.ResourceImporter{
+			StateContext: schema.ImportStatePassthroughContext,
+		},
+	}
+}
+
+// ReadDatabaseRole implements schema.ReadFunc.
+func ReadDatabaseRole(d *schema.ResourceData, meta interface{}) error {
+	db := meta.(*sql.DB)
+	dbRoleID, err := databaseRoleIDFromString(d.Id())
+	if err != nil {
+		return err
+	}
+
+	database := dbRoleID.DatabaseName
+	name := dbRoleID.RoleName
+
+	builder := snowflake.NewDatabaseRoleBuilder(name, database)
+	qry := builder.Show()
+	row := snowflake.QueryRow(db, qry)
+	// FIXME scan for name as there is LIKE pattern syntax
+	databaseRole, err := snowflake.ScanDatabaseRole(row)
+	databaseRole.DatabaseName = database // Nasty.
```

Comment:
not understanding what the point of this is. If database name is a ForceNew attribute then theres no reason to read  or overwrite it at all since its never going to change.